### PR TITLE
gh-61162: Clarify sqlite3 connection context manager docs

### PR DESCRIPTION
**Pull request by erlend-aasland**
Explicitly note that transactions are only closed if there is an open
transation at `__exit__`, and that transactions are not implicitly
opened during `__enter__`.

Co-authored-by: CAM Gerlach <CAM.Gerlach@Gerlach.CAM>
Co-authored-by: Stanley <46876382+slateny@users.noreply.github.com>

Automerge-Triggered-By: GH:erlend-aasland

### CODE REVIEW

**Comment by erlend-aasland** [comment]
cc. @python/proofreaders 

@Kodiologist, would you like to review this?

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [inline Doc/library/sqlite3.rst]:
```suggestion
    The context manager does not implicitly open a new transaction.
```

**Comment by AA-Turner** [comment]
I can't do an inline review as lines have been deleted and added. I'd suggest though focussing on the happy path of an open transaction and a successful body first, i.e.:

```rest
With an open transaction, connection objects can be used as context managers
that automatically commit or rollback transactions. If the body of the ``with``
statement finishes without exceptions, the transaction is committed. If an
exception is raised and not caught, the transaction is rolled back.

If there is no open transaction, the context manager is a no-op.
```

A

**Comment by erlend-aasland** [comment]
> I can't do an inline review as lines have been deleted and added. I'd suggest though focussing on the happy path of an open transaction and a successful body first, i.e.:

Thanks, that's an improvement!

**Review by AA-Turner — APPROVED** [review]
✅

A

**Comment by Kodiologist** [comment]
> @Kodiologist, would you like to review this?

I'm not a real CPython maintainer, but this looks good to me. Thanks.

**Comment by erlend-aasland** [comment]
I'll let this PR sit around a day or two before merging.

File: Doc/library/sqlite3.rst
Comment (mwichmann) [inline Doc/library/sqlite3.rst]: Mark this up as :keyword:`with`

File: Doc/library/sqlite3.rst
Comment (bskinn) [inline Doc/library/sqlite3.rst]: This sentence suggests that, even if a transaction is opened within the managed context, if there is no pre-existing open transaction at the time of context entry, then the context manager is a no-op. Is this right?

For example, in the [current example in the docs](https://docs.python.org/3/library/sqlite3.html?highlight=sqlite#using-the-connection-as-a-context-manager), would the first context manager be a no-op if the preceding `con.execute(...)` were immediate committed?

```
import sqlite3

con = sqlite3.connect(":memory:")
con.execute("create table lang (id integer primary key, name varchar unique)")
con.commit()  # NEW, EXPLICIT COMMIT PRIOR TO THE CONTEXT MANAGED USE OF con

# Successful, con.commit() is called automatically afterwards
#  !! Does the extra `con.commit()` now make this a no-op? !!
with con:
    con.execute("insert into lang(name) values (?)", ("Python",))

# con.rollback() is called after the with block finishes with an exception, the
# exception is still raised and must be caught
try:
    # !! Is this context manager a no-op unless there's already an open transaction? !!
    with con:
        con.execute("insert into lang(name) values (?)", ("Python",))
except sqlite3.IntegrityError:
    print("couldn't add Python twice")

# Connection object used as context manager only commits or rollbacks transactions,
# so the connection object should be closed manually
con.close()

```

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: No. Thanks for noticing. IIRC, my original wording was more correct.

Comment (AA-Turner) [reply Doc/library/sqlite3.rst]: Ahh, I read the original wording as "if there is an open transaction *at the point of calling `__enter__`*. Is the behaviour actually if there is an open transaction at the point of calling `__exit__`? 

A

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: > Ahh, I read the original wording as "if there is an open transaction *at the point of calling `__enter__`*. Is the behaviour actually if there is an open transaction at the point of calling `__exit__`? 

Yes. I'll get to it when I'm back at my laptop.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: We should consider noting explicitly that no transaction is started on enter. Actually, nothing happens on enter.

File: Doc/library/sqlite3.rst
Comment (bskinn) [inline Doc/library/sqlite3.rst:1438]: "... that automatically commit or rollback transactions that occur within the context." would probably be the most precise statement of this (if I understand it correctly).

----

That said, further clarification question: if there is an uncommitted transaction `t1` on `con` prior to entering the context (as in the example in the docs), what happens to it at `__exit__`? Based on my understanding and the various versions of the text, I would expect:

- If *no transactions* are made inside the context, then `t_1` remains on `con`, but in an uncommitted state (neither committed nor rolled back -- the context is a no-op)
- If one or more transactions `t_j` is added to `con` within the context *AND* if no uncaught error occurs in the context, I would expect `t1` to be committed along with all the `t_j` at context exit
- If one or more transactions `t_j` is added to `con` within the context *AND* an uncaught exception occurs, I would expect `t1` to be rolled back along with all of the `t_j` at context exit.

Regardless of whether the above is correct, I think some mention needs to be made about the disposition of pre-existing transactions on `con`.

Comment (bskinn) [reply Doc/library/sqlite3.rst:1438]: If what actually happens in the following example is that `t_1` is committed at `__exit__`, then the docs *definitely* need to call that out:

```
con = sqlite3.connect(...)
con.execute(sql_1)

with con:
    pass

# Is sql_1 now committed at this point?
```

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1438]: I'll try to make this explicit, technically:

- `__enter__`: nothing happens; returns self
- `__exit__`: if a transaction is open, commit or rollback; if commit failed, roll back; if no transaction is open, nothing happens

Now, we need to write this in a userfriendly way, trying to put emphasis on "the happy path".

Comment (AA-Turner) [reply Doc/library/sqlite3.rst:1438]: Does this fit the semantics?
```rest
Connection objects can be used as context managers that automatically commit or
rollback open transactions when leaving the body of the context manager.
If the body of the :keyword:`with` statement finishes without exceptions,
the transaction is committed.
If an exception is raised and not caught, the transaction is rolled back.

If there is no open transaction on leaving the body of the :keyword:`with`
statement, the context manager is a no-op.
```

A

Comment (bskinn) [reply Doc/library/sqlite3.rst:1438]: Recommend revision to the second half of the first paragraph (if correct to semantics):

"... finishes without exceptions, all open transactions on `con` are committed, *including* any transactions opened prior to entering the body of the context manager. If an exception is raised and not caught, all open transactions (including any created prior to the context manager) are rolled back."

Comment (AA-Turner) [reply Doc/library/sqlite3.rst:1438]: I'll let Erlend combine your suggested edits as he actually knows the semantics, I am just guessing!

A

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1438]: Thanks for your suggestions all. I think we should avoid being too verbose. Expect an update tonight (CET).

Comment (mwichmann) [reply Doc/library/sqlite3.rst:1438]: This feels convoluted, somehow.  The exit processing is to call commit (unless there was an uncaught exception), and that's already well-defined - it flushes out everything since the last commit, so I'm not sure we need the comments about no-op, or inside vs outside the with statement.

Without a great deal of thought, here's a different take on it, by using references to avoid getting too wordy.  Most context managers seem to close their object, so it seems worth mentioning that doesn't happen here.  

> A Connection object can be used as a context manager,
typically to wrap a transaction.
Upon successful completion of the :keyword:`with` block,
a :meth:`commit` is performed.
In case of a failed commit or an uncaught exception,
a :meth:`rollback` is performed.
The connection is not closed.

However, I don't know if this actually addresses the original complaint in #61162.

Comment (bskinn) [reply Doc/library/sqlite3.rst:1438]: I'm for sure hung up on the ambiguity of what happens to preexisting transactions. Absolutely, that could be a total side issue.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1438]: > The exit processing is to call commit (unless there was an uncaught exception), and that's already well-defined [...]

But that is not the reality (!), and we cannot change the status quo without breaking people's code. What we _can_ do, however, is document the actual behaviour, to get rid of the confusion, so that is what we're trying to do here :)

I'll repeat what I said earlier in this thread, a little bit more verbose:

`__enter__` does nothing ... the current confusion starts already here, because a lot of people _think_ that `__enter__` opens a new transaction! But it does not, and assuming otherwise can bite your code, and you can end up debugging your code for days without finding out what is wrong. Take a look at the linked issue, and you'll see the complaints ;) This is why _I_ think it is a good idea to document, and maybe even add emphasis to, the fact that _no transaction is opened when you enter the context manager_. The original commit of this PR had a `note` directive for this fact.

`__exit__`; this is the source of the rest of the confusion. A lot of people _think_ that the context manager will commit/rollback at exit. But that is not the case. Commit/rollback is only performed _if there is an open transaction_. Since the context manager (confusingly) does _not_ open a transaction, but people assume so, we get complains, because people experience unexpected behaviour. Now, we _can_ do something with the expectations by improving the docs ;) In the included example in the docs, there is an implicit transaction opened _because a DML query is executed during the with body_, so at `__exit__`, we can expect commit or rollback.

As you'll see from the issue, the original pysqlite author (Gerhard) is +1 to deprecating the context manager, and I fully understand that :)

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1438]: All right, PTAL! :)

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1438]: And to add to the confusion: if `isolation_level` is not `None` and you are executing a DML[^1] statement, a transaction is implicitly started, for `execute` and `executemany`, but not for `executescript`.

[^1]: INSERT, UPDATE, DELETE, and REPLACE. But not CREATE TABLE, for example.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1438]: Brian, to briefly answer your https://github.com/python/cpython/pull/93890#discussion_r899130035:

It depends ;) It depends on the value of `isolation_level`, and it depends on the type of query you are executing. Now, we're trying to reduce all the "it depends" by introducing a new way of [wielding transaction control](https://discuss.python.org/t/introducing-a-pep-249-compliant-transaction-behaviour-to-sqlite3/16516?u=erlendaasland), and as I'm hinting to in the linked issue of this PR, possibly introducing a more consistent context manager based on SAVEPOINT/RELEASE.

Hopefully, sqlite3 in Python 3.14 will be less confusing :)

Comment (mwichmann) [reply Doc/library/sqlite3.rst:1438]: > > The exit processing is to call commit (unless there was an uncaught exception), and that's already well-defined [...]
> 
> But that is not the reality (!), and we cannot change the status quo without breaking people's code. 

I assume I'm missing something (outer wrapping around this maybe?), but this is what I see in the sqlite module code, which looks that *is* what it does:

```C
static PyObject *
pysqlite_connection_exit_impl(pysqlite_Connection *self, PyObject *exc_type,
                              PyObject *exc_value, PyObject *exc_tb)
/*[clinic end generated code: output=0705200e9321202a input=bd66f1532c9c54a7]*/
{
    int commit = 0;
    PyObject* result;

    if (exc_type == Py_None && exc_value == Py_None && exc_tb == Py_None) {
        commit = 1;
        result = pysqlite_connection_commit_impl(self);
    }
    else {
        result = pysqlite_connection_rollback_impl(self);
    }
    ...
```

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1438]: You _assume_ what those functions are doing. Now, follow those function calls one level further.

<details>
<summary>Spoilers</summary>

https://github.com/python/cpython/blob/f8e576be0a7cd38f753f31cf4178db81a602fc32/Modules/_sqlite/connection.c#L529-L555

https://github.com/python/cpython/blob/f8e576be0a7cd38f753f31cf4178db81a602fc32/Modules/_sqlite/connection.c#L564-L591

The key in both of those functions is `if (!sqlite3_get_autocommit(self->db)) {`. Now, what does `sqlite3_get_autocommit` do? It [tests for SQLite's autocommit mode](https://www.sqlite.org/c3ref/get_autocommit.html):

> The sqlite3_get_autocommit() interface returns non-zero or zero if the given database connection is or is not in autocommit mode, respectively. Autocommit mode is on by default. Autocommit mode is disabled by a [BEGIN](https://www.sqlite.org/lang_transaction.html) statement. Autocommit mode is re-enabled by a [COMMIT](https://www.sqlite.org/lang_transaction.html) or [ROLLBACK](https://www.sqlite.org/lang_transaction.html).

We know that BEGIN is shorthand for BEGIN TRANSACTION, so BEGIN opens a transaction. This implies that `sqlite3_get_autocommit()` returns zero when a transaction is open. Negate that return value and you've got a truthy value for our key `if` statement. So, we end up with "if in transaction, commit" and "if in transaction, rollback". There is no `else`, so the "if not in transaction" passes silently.

</details>

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst:1436]:
```suggestion
.. _sqlite3-connection-context-manager:

Using the connection as a context manager
```

Add this here to link in NEWS

File: Misc/NEWS.d/next/Documentation/2022-06-16-10-10-59.gh-issue-61162.1ypkG8.rst
Comment (CAM-Gerlach) [inline Misc/NEWS.d/next/Documentation/2022-06-16-10-10-59.gh-issue-61162.1ypkG8.rst]: For clarity, I'd think it would be worth adding a link here to the actual section added.

```suggestion
Clarify :mod:`sqlite3` behavior when :ref:`sqlite3-connection-context-manager`.
```

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
:class:`Connection` objects can be used as context managers that automatically commit or
```

Give it a proper role and a link, since it is described over 1000 lines above.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst:1443]:
```suggestion
and there is an open transaction
(including if opened prior to entering the context manager),
it is committed.
```

Would this help clarify, given the discussion above?

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
If this commit fails,
or the body of the ``with`` statement raise an exception which is not caught,
```

Make more concise, less choppy/repetitive, and clearer that the same action is taken on either condition without the reader having to carefully inspect both (since they might expect it not to be, as putting both in the same sentence would typically be the more idiomatic pattern here, so having them be in different ones might lead the reader to expect them to be different).

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
   The context manager does not implicitly open a new transaction
   or close the connection.
```

Likewise, same rationale as above.

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1438]: Even better, you can test the functionality using the sqlite3 extension module. The connection attribute `in_transaction` is mapped _directly_ to the _inverted_ result of `sqlite3_get_autocommit`.

<details>
<summary>Test: open transaction before ctx mgr</summary>

```py
>>> import sqlite3
>>> cx = sqlite3.connect(":memory:")
>>> cx.set_trace_callback(print)
>>> cx.in_transaction
False
>>> cx.execute("BEGIN")  # explicitly opening a transaction
BEGIN
<sqlite3.Cursor object at 0x7fd85413aa40>
>>> cx.in_transaction
True
>>> with cx: pass
...
COMMIT
>>> cx.in_transaction
False
```

</details>

<details>
<summary>Test: no transaction before ctx mgr</summary>

```py
>>> import sqlite3
>>> cx = sqlite3.connect(":memory:")
>>> cx.set_trace_callback(print)
>>> cx.in_transaction
False
>>> with cx: pass
...
>>> cx.in_transaction
False
```

</details>

<details>
<summary>Test: implicit transaction opened during ctx mgr</summary>

```py
>>> import sqlite3
>>> cx = sqlite3.connect(":memory:")
>>> cx.set_trace_callback(print)
>>> cx.execute("create table t(t)")
create table t(t)
<sqlite3.Cursor object at 0x7f4a06d72a40>
>>> cx.in_transaction
False
>>> with cx: cx.execute("insert into t values(1)")
...
BEGIN
insert into t values(1)
<sqlite3.Cursor object at 0x7f4a06d72a40>
COMMIT
>>> cx.in_transaction
False
```

</details>

<details>
<summary>Test: non-DML statement in with body</summary>

```py
>>> import sqlite3
>>> cx = sqlite3.connect(":memory:")
>>> cx.set_trace_callback(print)
>>> cx.in_transaction
False
>>> with cx: cx.execute("create table t(t)")
...
create table t(t)
<sqlite3.Cursor object at 0x7fd186825730>
>>> cx.in_transaction
False
```

</details>

File: Doc/library/sqlite3.rst
Comment (mwichmann) [inline Doc/library/sqlite3.rst:1446]: Perhaps combine these, e.g. "If the commit fails, or if there was an uncaught exception in the body, the transaction is rolled back".

**Review by CAM-Gerlach — COMMENTED** [review]
Some Sphinx and writing suggestions

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1436]: Ah, neat. Thanks.

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1443]: I'm not sure. BEGIN ... COMMIT cannot nest, so there can only be one transaction for each connection. In theory, you could have layers of SAVEPOINT's inside a BEGIN ... COMMIT. I'm afraid such a wording may end up causing more confusion for the average user.

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1446]: See 8808fb91b0d12321ee1e3a408e81c708ad8dbf90

File: Doc/library/sqlite3.rst
Comment (mwichmann) [inline Doc/library/sqlite3.rst]: This is just personal nitpickery, but I think this is better as a singular: "A Connection object....as a context manager".

Question: the docs say that you'll get a connection object *unless* you supply a factory (See `connect` method doc). My inclination would be to not elaborate on that later, and take as suggested (`:class:Connection`)

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1438]: I consider this conversation resolved. If there are more questions, please start a new conversation.

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Yes, both connections and cursors can be created using factories. It is required, of course, that the original classes are used as some kind of parent / base. I don't think we need to mention this here; it will not make the docs more clear.

File: Doc/library/sqlite3.rst
Comment (mwichmann) [reply Doc/library/sqlite3.rst:1438]: > You _assume_ what those functions are doing. Now, follow those function calls one level further.

So (real question, no criticism): does that mean there ought to be more detail in the commit/rollback descriptions, rather than the context manager bit?

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1438]: The DB API has the following specification for commit and rollback. Relevant sentences quoted:

> Commit any pending transaction to the database.

> In case a database does provide transactions this method causes the database to roll back to the start of any pending transaction.

The docs seems to reflect that pretty good, but perhaps the docstrings could be made more accurate. That is out of scope of this PR, but I think there is room for improvement.

**UPDATE**: I created #93925

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1443]: I'll sleep on it for a day or two. Right now, I'm leaning towards singular.

Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst:1443]: Oh, I wasn't sure about that, thanks—the main intent was to clarify the point about a transaction opened prior to entering the contextmanager. I edited the suggestion above to clarify that point and use the singular.

**Review by ezio-melotti — APPROVED** [review]

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1443]: I'm afraid it becomes to verbose. The sentence three lines below reads:

> If there is no open transaction upon leaving the body of the ``with`` statement, the context manager is a no-op.

The preceding sentence reads:

> A :class:`Connection` object can be used as a context manager that automatically commit or rollback open transactions when leaving the body of the context manager.

Isn't that enough?

File: Doc/library/sqlite3.rst
Comment (slateny) [inline Doc/library/sqlite3.rst]:
```suggestion
automatically commits or rollbacks open transactions when leaving the body of the
```
I think the 's' is needed since object is singular

File: Doc/library/sqlite3.rst
Comment (slateny) [inline Doc/library/sqlite3.rst]: similarly:

```suggestion
or the body of the ``with`` statement raises an exception which is not caught,
```

or optionally:

```suggestion
or if the body of the ``with`` statement raises an uncaught exception,
```

File: Doc/library/sqlite3.rst
Comment (pradyunsg) [reply Doc/library/sqlite3.rst:1443]: I reckon that the current form is good enough. :)

File: Doc/library/sqlite3.rst
Comment (bskinn) [reply Doc/library/sqlite3.rst]: Subtlety here... The noun for the action is a 'rollback', but the verb for it is 'to roll back' (two words). So, this should be 'automatically commits or rolls back'.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]:
```suggestion
automatically commits or rolls back open transactions when leaving the body of the
```

**Comment by erlend-aasland** [comment]
I'll say it again, and see what happens: I'll let this PR sit around for a couple of days more before merging :)

Thanks, Proofreaders 🪄

File: Doc/library/sqlite3.rst
Comment (pradyunsg) [inline Doc/library/sqlite3.rst:1442]:
```suggestion
If the body of the :keyword:`with` statement finishes without an uncaught exception,
```

*Maybe*?

**Review by pradyunsg — APPROVED** [review]

**Review by ezio-melotti — APPROVED** [review]

**Review by CAM-Gerlach — APPROVED** [review]
LGTM, thanks @erlend-aasland !

**Pull request merged**
